v2/pkg/engine/datasource/graphql_datasource/graphql_subscription_client_test.go

```diff
@@ -1883,3 +1886,457 @@ func TestAsyncSubscribe(t *testing.T) {
+			assert.NoError(t, err)
+
+			// Send ping message
+			err = conn.Write(r.Context(), websocket.MessageText, []byte(`{"type":"ping"}`))
+			assert.NoError(t, err)
+
+			// Wait for pong response from client
+			msgType, data, err = conn.Read(ctx)
+			if err != nil {
+				t.Errorf("Error reading pong: %v", err)
+				return
+			}
+
+			if string(data) == `{"type":"pong"}` {
+				assert.Equal(t, websocket.MessageText, msgType)
+				close(pongReceived)
+
+				// Send another data message
+				err = conn.Write(r.Context(), websocket.MessageText, []byte(`{"id":"1","type":"next","payload":{"data":{"messageAdded":{"text":"after ping-pong"}}}}`))
+				assert.NoError(t, err)
+			}
+
+			// Wait for client to unsubscribe (complete message)
+			readTimeout := time.NewTimer(3 * time.Second)
+			defer readTimeout.Stop()
+
+			readDone := make(chan struct{})
+			go func() {
+				msgType, data, _ = conn.Read(ctx)
+				close(readDone)
+			}()
+
+			select {
+			case <-readDone:
+				// Successfully read client message
+			case <-readTimeout.C:
+				// Timeout is fine, we're just waiting for unsubscribe
+			}
+
+			close(serverDone)
+		}))
+		defer server.Close()
+
+		ctx, clientCancel := context.WithCancel(context.Background())
+		defer clientCancel()
+		serverCtx, serverCancel := context.WithCancel(context.Background())
+		defer serverCancel()
+
+		client := NewGraphQLSubscriptionClient(http.DefaultClient, http.DefaultClient, serverCtx,
+			WithLogger(logger()),
+			WithReadTimeout(100*time.Millisecond),
+			WithNetPollConfiguration(NetPollConfiguration{
+				Enable:           true,
+				TickInterval:     100 * time.Millisecond,
+				BufferSize:       10,
+				MaxEventWorkers:  2,
+				WaitForNumEvents: 1,
+			}),
+		).(*subscriptionClient)
+
+		updater := &testSubscriptionUpdater{}
+
+		err := client.SubscribeAsync(resolve.NewContext(ctx), 2, GraphQLSubscriptionOptions{
+			URL: server.URL,
+			Body: GraphQLBody{
+				Query: `subscription {messageAdded(roomName: "room"){text}}`,
+			},
+			WsSubProtocol: ProtocolGraphQLTWS,
+		}, updater)
+		assert.NoError(t, err)
+
+		// Wait some time to allow subscription processing
+		time.Sleep(2 * time.Second)
+
+		// Verify we receive at least the initial data
+		updater.mux.Lock()
+		updatesCount := len(updater.updates)
+		firstUpdate := ""
+		if updatesCount > 0 {
+			firstUpdate = updater.updates[0]
+		}
+		updater.mux.Unlock()
+
+		assert.GreaterOrEqual(t, updatesCount, 1)
+		if updatesCount > 0 {
+			assert.Equal(t, `{"data":{"messageAdded":{"text":"initial data"}}}`, firstUpdate)
+		}
+
+		// Cleanup
+		client.Unsubscribe(2)
+
+		// Wait for server to finish
+		select {
+		case <-serverDone:
+			// Server completed successfully
+		case <-time.After(5 * time.Second):
+			t.Fatal("Timed out waiting for server to complete")
+		}
+
+		clientCancel()
+		serverCancel()
+	})
+}
+
+func TestClientClosesConnectionOnPingTimeout(t *testing.T) {
```

Comment:
Can you the following tests:
1. server sends pong every 100ms, independent of other workloads
2. server sends pong before responding with ack
3. protocol ws example where server sends "ka" events

Side note. How is the "ping" scenario handled with graphql-ws?
There's no ping, just "ka" from server to client (engine is client).
In this case, how do we know if a server is alive?